File: Adopt-A-Dog/artifacts/adopt-a-dog/services/dogApiService.ts
export interface Breed {
  name: string;
  subBreeds: string[];
}

const BASE = "https://dog.ceo/api";

export async function fetchBreeds(): Promise<Breed[]> {
  const res = await fetch(`${BASE}/breeds/list/all`);
  if (!res.ok) throw new Error("Failed to load breeds");
  const data = await res.json();
  const message: Record<string, string[]> = data.message;
  return Object.entries(message)
    .map(([name, subBreeds]) => ({ name, subBreeds }))
    .sort((a, b) => a.name.localeCompare(b.name));
}

export async function fetchRandomImage(breedPath: string): Promise<string> {
  const res = await fetch(`${BASE}/breed/${breedPath}/images/random`);
  if (!res.ok) throw new Error("Failed to load image");
  const data = await res.json();
  return data.message as string;
}

export async function fetchMultipleImages(
  breedPath: string,
  count: number = 4
): Promise<string[]> {
  const res = await fetch(`${BASE}/breed/${breedPath}/images/random/${count}`);
  if (!res.ok) throw new Error("Failed to load images");
  const data = await res.json();
  return data.message as string[];
}

export async function fetchRandomBreedAndImage(): Promise<{
  breed: string;
  imageUrl: string;
}> {
  const breeds = await fetchBreeds();
  const random = breeds[Math.floor(Math.random() * breeds.length)];
  const imageUrl = await fetchRandomImage(random.name);
  return { breed: random.name, imageUrl };
}

File: lib/services/dog_api_service.dart
import 'dart:convert';
import 'package:adopt_a_dog/models/breed.dart';
import 'package:http/http.dart' as http;

class DogApiService {
  static const _base = 'https://dog.ceo/api';

  Future<List<Breed>> fetchBreeds() async {
    final uri = Uri.parse('$_base/breeds/list/all');
    final response = await http.get(uri);

    if (response.statusCode != 200) {
      throw Exception('Failed to load breeds');
    }

    final Map<String, dynamic> data = jsonDecode(response.body);
    final Map<String, dynamic> message = data['message'];

    return message.entries
        .map(
          (e) =>
              Breed(name: e.key, subBreeds: List<String>.from(e.value as List)),
        )
        .toList()
      ..sort((a, b) => a.name.compareTo(b.name));
  }

  /// Fetches a random image URL for the given breed (and optional sub-breed).
  Future<String?> fetchRandomImage(Breed breed, {String? sub}) async {
    final path = breed.apiPath(sub: sub);
    final uri = Uri.parse('$_base/$path/images/random');
    final response = await http.get(uri);

    if (response.statusCode != 200) {
      return null;
    }

    final Map<String, dynamic> data = jsonDecode(response.body);
    return data['message'] as String;
  }

  /// Fetches multiple random images for a breed.
  Future<List<String>> fetchImages(Breed breed, {String? sub, int count = 6}) async {
    final path = breed.apiPath(sub: sub);
    final uri = Uri.parse('$_base/$path/images/random/$count');
    final response = await http.get(uri);

    if (response.statusCode != 200) {
      return [];
    }

    final Map<String, dynamic> data = jsonDecode(response.body);
    return List<String>.from(data['message'] as List);
  }
}
